numpy/linalg/umath_linalg.cpp

```diff
@@ -983,11 +983,24 @@ identity_matrix(typ *matrix, size_t n)
     }
 }
 
-         /* lower/upper triangular matrix using blas (in place) */
+         /* zero the undefined part in a upper/lower triangular matrix */
 
 template<typename typ>
 static inline void
 triu_matrix(typ *matrix, size_t n)
+{
+    size_t i, j;
+    for (i = 0; i < n-1; ++i) {
+        for (j = i+1; j < n; ++j) {
+            matrix[j] = numeric_limits<typ>::zero;
+        }
+        matrix += n;
+    }
+}
```

Comment:
I've got a question: I see that `triu_matrix` got renamed to `tril_matrix`. Was it incorrect?
AFAIU `triu_matrix` zeroes the upper part of the matrix instead of the lower (assuming row-major-order), or is it column-major implementation?

Reply:
This is just what git diff does - really what happened here is that I added `triu_matrix` - `tril_matrix continues to live below!

Reply:
Here's an implementation of `triu_matrix` in the current `main` branch: 
https://github.com/numpy/numpy/blob/7b42b5242fb8e9c2b9636ebdd8720299ac37b5f8/numpy/linalg/umath_linalg.cpp#L987-L1001

And here's `triu_matrix` from your branch:
https://github.com/mhvk/numpy/blob/502bfb27968746f70155130b5fa9579e4d9a3a85/numpy/linalg/umath_linalg.cpp#L987-L1000

It looks like the implementation of `triu_matrix` changed here, right?

Reply:
What I mean: `tril_matrix` is the one that is new and `triu_matrix` is the existing one, and I wonder why the existing one changed implementation. I might be missing something.

Reply:
I think this is a typo in the `main` branch version. In `def cholesky` there's `assert(uplo == 'L');` but later `triu_matrix(params.A, params.N);` is called. 
So I think the change here is correct, but the release note should say that `triu_matrix` implementation changed and `tril_matrix` was added.

Reply:
Ah, yes, sorry! I did change the implementation, as indeed for a triangular-upper matrix, one should be zeroing the lower part. I now recall puzzling over this too, and changing the comment above to try to make clear what was actually done.

Reply:
Ok! I've got one more question: Here `triu_matrix` is zeroing upper triangular if one assumes the input is row-major order and lower triangular if one assumes the input column-major order, right?

Then maybe we could have just a docstring that explains why:
```python
triu_matrix(typ *matrix, size_t n)
{
    size_t i, j;
    for (i = 0; i < n-1; ++i) {
        for (j = i+1; j < n; ++j) {
            matrix[j] = numeric_limits<typ>::zero;
        }
        matrix += n;
    }
}
```
looks like it zeroes upper triangular but the function's name says `triu`? Just a comment that `matrix` is column major.
WDYT?

Reply:
Yes, makes sense!

Reply:
OK, added the extra comment (really good to have indeed!)